clang/lib/Sema/SemaExpr.cpp

```diff
@@ -18931,7 +18931,7 @@ void Sema::MarkFunctionReferenced(SourceLocation Loc, FunctionDecl *Func,
   //   constant evaluated
   bool NeededForConstantEvaluation =
       isPotentiallyConstantEvaluatedContext(*this) &&
-      isImplicitlyDefinableConstexprFunction(Func);
+      isImplicitlyDefinableConstexprFunction(Func) && !Func->isPure();
```

Comment:
I don't understand how the function being pure relates to whether it's needed for constant evaluation. I think this should move down to `NeedDefinition` below?

Reply:
I concur, this does not make sense doing the check here.

Reply:
I am not sure the diagnostic makes sense in that case since it can never actually be called.

Reply:
Thank to @erichkeane I realized I am wrong, we can actually define a pure virtual function. So we do want to retain the diagnostic here.

Reply:
I am not sure why moving it down silences the diagnostic, I would think doing `(OdrUse == OdrUseContext::Used || NeededForConstantEvaluation || !Func->isPure())` should work. 

Can you explain in more detail?

Reply:
You're actually right. 

I don't really recall what I was entering back when I was developing it, but it should work now.